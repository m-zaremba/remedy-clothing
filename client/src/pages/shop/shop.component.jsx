import React, { useEffect } from 'react';
import { Route } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import CollectionsOverviewContainer from '../../components/collections-overview/collections-overview.container';
import CollectionPageContainer from '../collection/collection.container';

import { fetchCollectionsStart } from '../../redux/shop/shop.actions';

const ShopPage = ({ fetchCollectionsStart, match }) => {
  useEffect(() => {
    fetchCollectionsStart();
  }, [fetchCollectionsStart]);

  return (
    <div>
      <Route exact path={`${match.path}`} component={CollectionsOverviewContainer} />
      <Route path={`${match.path}/:collectionId`} component={CollectionPageContainer} />
    </div>
  );
};

const mapDispatchToProps = (dispatch) => ({
  fetchCollectionsStart: () => dispatch(fetchCollectionsStart()),
});

ShopPage.propTypes = {
  match: PropTypes.shape({
    path: PropTypes.string,
    url: PropTypes.string,
    isExact: PropTypes.bool,
    params: PropTypes.object,
  }).isRequired,
  fetchCollectionsStart: PropTypes.func.isRequired,
};

export default connect(null, mapDispatchToProps)(ShopPage);
