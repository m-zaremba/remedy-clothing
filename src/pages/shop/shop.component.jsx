import React, { useEffect, useContext, useState } from "react";
import { Route } from "react-router-dom";

import {
  firestore,
  convertCollectionsSnapshotToMap,
} from "../../firebase/firebase.utils";

import { CollectionsContext } from "../../providers/collections/collections.provider";

import CollectionsOverview from "../../components/collections-overview/collections-overview.component";
import CollectionPage from "../collection/collection.component";

const ShopPage = ({ match }) => {
  const { collections, setContextCollections } = useContext(CollectionsContext);
  const [fetchedCollections, setFetchedCollections] = useState(null);

  useEffect(() => {
    const unsubscribe = firestore
      .collection("collections")
      .onSnapshot(async (snapshot) => {
        const convertedCollections = convertCollectionsSnapshotToMap(snapshot);
        setFetchedCollections(convertedCollections);
      });
    return () => unsubscribe();
  }, []);

  useEffect(() => {
    setContextCollections(fetchedCollections);
  }, [fetchedCollections, setContextCollections]);

  return (
    collections && (
      <div className="shop-page">
        <Route exact path={`${match.path}`} component={CollectionsOverview} />
        <Route
          path={`${match.path}/:collectionId`}
          component={CollectionPage}
        />
      </div>
    )
  );
};

export default ShopPage;
