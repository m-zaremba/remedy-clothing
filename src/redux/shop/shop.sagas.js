import { takeLatest, call, put } from 'redux-saga/effects';

import { firestore, convertCollectionsSnapshotToMap } from '../../firebase/firebase.utils';

import { fetchCollectionSuccess, fetchCollectionsFailure } from './shop.actions';

import shopActionTypes from './shop.types';

export function* fetchCollectionsAsync() {
  try {
    const collectionRef = firestore.collection('collections');
    const snapshot = yield collectionRef.get();
    const collectionsMap = yield call(convertCollectionsSnapshotToMap, snapshot);
    yield put(fetchCollectionSuccess(collectionsMap));
  } catch (error) {
    yield put(fetchCollectionsFailure(error.message));
  }
}

export function* fetchCollectionsStart() {
  yield takeLatest(shopActionTypes.FETCH_COLLECTIONS_START, fetchCollectionsAsync);
}
