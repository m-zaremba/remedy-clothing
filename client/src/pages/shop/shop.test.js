import React from 'react';
import { mount } from 'enzyme';
import { combineReducers, createStore } from 'redux';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import { ShopPage } from './shop.component';

export const mockStore = ({ state, reducers }) => {
  const store = createStore(combineReducers(reducers), state);
  return {
    ...store,
  };
};

describe('ShopPage', () => {
  let wrapper;
  let mockFetchCollectionsStart;
  let store;

  beforeEach(() => {
    const mockReducer = (
      state = {
        isFetching: true,
      },
      action,
    ) => state;

    const mockState = {
      shop: {
        isFetching: true,
      },
    };

    mockFetchCollectionsStart = jest.fn();

    store = mockStore({
      state: mockState,
      reducers: { shop: mockReducer },
    });

    const mockMatch = {
      path: '',
    };

    const mockProps = {
      match: mockMatch,
      fetchCollectionsStart: mockFetchCollectionsStart,
    };

    wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <ShopPage {...mockProps} />
        </Provider>
      </BrowserRouter>,
    );
  });

  it('should render ShopPage component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call fetchCollectionStart', () => {
    expect(mockFetchCollectionsStart).toHaveBeenCalled();
  });
});
