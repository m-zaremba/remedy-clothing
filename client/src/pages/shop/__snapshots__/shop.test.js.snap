// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopPage should render ShopPage component 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <ShopPage
        fetchCollectionsStart={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        match={
          Object {
            "path": "",
          }
        }
      >
        <div>
          <Suspense
            fallback={<Spinner />}
          >
            <Route
              component={
                Object {
                  "$$typeof": Symbol(react.lazy),
                  "_ctor": [Function],
                  "_result": Object {
                    "$$typeof": Symbol(react.memo),
                    "WrappedComponent": [Function],
                    "compare": null,
                    "displayName": "Connect(Component)",
                    "type": [Function],
                  },
                  "_status": 1,
                }
              }
              exact={true}
              path=""
            />
            <Route
              component={
                Object {
                  "$$typeof": Symbol(react.lazy),
                  "_ctor": [Function],
                  "_result": null,
                  "_status": -1,
                }
              }
              path="/:collectionId"
            />
            <Spinner>
              <styled.div>
                <div
                  className="sc-AxjAm HvuBM"
                >
                  <styled.div>
                    <div
                      className="sc-AxirZ haZpvb"
                    />
                  </styled.div>
                </div>
              </styled.div>
            </Spinner>
          </Suspense>
        </div>
      </ShopPage>
    </Provider>
  </Router>
</BrowserRouter>
`;
